TRT 4.0 update

jjsjann123 opened the pull request:
updated feature support for TRT 4.0 layers
  disabled broken shape inference (added TODO)
  removed unused code
  code is compatible with TRT 3.0.4

jjsjann123 committed:
TRT 4.0 update
  code is compatible with TRT 3.0.4
  updated feature support for TRT 4.0 layers
  added error checking for converter
  disabled broken shape inference (added TODO)

jjsjann123 commented:
Ping @aaroey @benbarsdell @samikama

aaroey reviewed (changes requested):
High level comment: please add unit tests to verify that each converter is doing the right thing. I know this is hard, especially when there are no tests done before. But it would be even harder later if we don't do it today, and the lack of test will make future development break things easily. So, please follow the convention and add tests.

These are just initial comments, more will come.
Thanks.

aaroey commented:
@jjsjann123 Please add the tests in this PR then, I'm happy to review them at the same time.

jjsjann123 commented:
even better! Will push it later today.
just FYI, unit tests are totally separate from the code so there's no need to hold back the review waiting for my push

really appreciate the review :)

jjsjann123 committed:
addressing review comments
added Constant broadcast in converter;
added integration tests for converter;

jjsjann123 commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -53,6 +53,31 @@ limitations under the License.
+#define TFTRT_RETURN_ERROR_IF_FALSE(ptr, node)                               \
+  do {                                                                       \
+    if (ptr == false) {                                                      \
+      return tensorflow::errors::Internal(string("TFTRT::") + __FUNCTION__ + \
```
Manually separated the lines.
But seems to be violating clang-format. (The automatic correction put the lines back)

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -75,13 +102,108 @@ inline tensorflow::Status ConvertDType(tensorflow::DataType tf_dtype,
+    if (max_d != l_d) {  // if broadcast beyond batch dimension, fail
+      return false;
+    }
+    l_s[0] = -1;
```
It seems this does not align with what the comment says:
```
+    T: 3 5 1
+    -> T: 1 1 1 -1 3 5 1
```
but here it assigns -1 to first element.

jjsjann123 replied:
The comment is trying to illustrate why we cannot support this case.
In the code, if we need to promote the rank of the tensor operand, the state if (max_d != l_d) returns true and the function returns false.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -790,118 +918,23 @@ tensorflow::Status BinaryCompute(const TRT_ShapedWeights& iweights_l,
     Converter& ctx, const tensorflow::NodeDef& node_def,
     const nvinfer1::ITensor* tensor, TRT_ShapedWeights weights,
-    std::vector<TRT_TensorOrWeights>* outputs) {
+    std::vector<TRT_TensorOrWeights>* outputs, bool swapped_inputs) {
```
Please follow the convention by making all input parameters appear before any output parameter.
Also, what does swapped_inputs means? Please comment.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -910,6 +943,12 @@ tensorflow::Status BinaryTensorOpWeight(
   auto dims_t = tensor->getDimensions();
 
+  // TODO(jie): addScale checks for input tensor dimension
+  if (dims_t.nbDims != 3) {
```
Can we put all checks like this into a method and call it in the segmenter, so we can avoid putting nodes into the trt engine at the very beginning, instead of delaying the check until here?

jjsjann123 replied:
We have undergoing refactoring work on the segmenter. It is to improve the robustness of the converter and expand checking only the op types to consider attributes as well as input types.

So that will come in as future PRs.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1056,6 +1123,17 @@ tensorflow::Status ConvertConv2DHelper(
+    VLOG(2) << weights_rsck.shape_.d[i];
+  }
+
+  if (weights_rsck.shape_.nbDims != 4) {
```
Same question about putting the checks in segmenter.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1147,18 +1232,71 @@ tensorflow::Status ConvertConv2DHelper(
                                            node_def.name());
 }
 
+bool PrepareTensorForShape(Converter& ctx, const TRT_TensorOrWeights& input,
```
Please fix the order of input and output parameters here and all other places in this PR.

jjsjann123 replied:
Did a git diff and fixed two other functions with the same problem :)

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1147,18 +1232,71 @@ tensorflow::Status ConvertConv2DHelper(
                                            node_def.name());
 }
 
+bool PrepareTensorForShape(Converter& ctx, const TRT_TensorOrWeights& input,
```
Please comment what this method does.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1147,18 +1232,71 @@ tensorflow::Status ConvertConv2DHelper(
 tensorflow::Status BinaryTensorOpTensor(
     Converter& ctx, const tensorflow::NodeDef& node_def,
-    const nvinfer1::ITensor* tensor_l, const nvinfer1::ITensor* tensor_r,
+    const TRT_TensorOrWeights operand_l, const TRT_TensorOrWeights operand_r,
```
const xx&?

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1468,39 +1608,39 @@ tensorflow::Status ConvertConst(Converter& ctx,
+      std::vector<float> tensor_data(
+          weights_tensor.float_val().begin(),
+          weights_tensor.float_val()
+              .end());  //  make a local copy first to flatten
```
Please put the comment on a separate line above the code line.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1520,11 +1660,17 @@ tensorflow::Status ConvertConst(Converter& ctx,
+      std::vector<int32> tensor_data(
+          weights_tensor.int_val().begin(),
+          weights_tensor.int_val()
+              .end());  //  make a local copy first to flatten
```
Same here.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1694,10 +1895,82 @@ tensorflow::Status ConvertReduce(Converter& ctx,
   outputs->push_back(TRT_TensorOrWeights(output_tensor));
   return tensorflow::Status::OK();
 }
+#endif
```
`#elseif NV_TENSORRT_MAJOR > 3`

jjsjann123 replied:
Good catch

aaroey commented on tensorflow/contrib/tensorrt/test/unit_tests/BatchMatMulTest.py:
```diff
@@ -0,0 +1,102 @@
+from tensorflow.python.ops import math_ops
+from tensorflow.python.layers import core
+from tensorflow.python.training import training
+from base_unit_test import BaseUnitTest
```
Thanks so much for adding the tests!

Please use absolute path and add corresponding build targets for the tests.
Same for other tests.

aaroey commented on tensorflow/contrib/tensorrt/test/unit_tests/BatchMatMulTest.py:
```diff
@@ -0,0 +1,102 @@
+          np.random.randn(5, 1, 1), dtype=dtypes.float32)
+      x1 = x1 + b
+
+      var = variable_scope.get_variable("test", [12, 5, 12, 7], dtype=dtypes.float32, initializer=init_ops.truncated_normal_initializer)
```
Please fix the formatting for all the tests.

jjsjann123 commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -75,13 +102,108 @@ inline tensorflow::Status ConvertDType(tensorflow::DataType tf_dtype,
+
+  int max_d = std::max(l_d, r_d);
+  std::memcpy(l_s + max_d - operand_l.nbDims, operand_l.d,
+              operand_l.nbDims * element_size);
```
Yes this is true. Do we want to replace element_size with sizeof(int)?

jjsjann123 committed:
[addressing review comments]
  i. correcting code style
  2. adding comments on functions

jjsjann123 committed:
[addressing review comments]
  1. addressing comments on coding style
  2. addressing comments on macro logics
  3. using absolute path for python tests
  4. added building targets for unit tests
  5. formating python scripts (using yapf with google style + 2 spaces indent)

aaroey committed:
Fix shape fn

aaroey committed:
Add TODOs and fix some formatting issues

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -49,9 +49,36 @@ limitations under the License.
+
+#define TFTRT_RETURN_ERROR_IF_FALSE(ptr, node) \
+  do {                                         \
+    if (ptr == false) {                        \
```
s/false/nullptr? Is ptr a pointer? If so what's the difference between this macro and the xx_IF_NULLPTR below? Otherwise please rename 'ptr' to something else.

jjsjann123 replied:
I'll rename ptr.

jjsjann123 commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1567,47 +1712,105 @@ tensorflow::Status ConvertBinary(Converter& ctx,
+    auto status = BinaryTensorOpWeight(ctx, node_def, inputs.at(0).tensor(),
+                                       inputs.at(1).weights(), false, outputs);
+#if NV_TENSORRT_MAJOR == 3
+    TF_RETURN_IF_ERROR(status);
```
For performance consideration, we prefer to use BinaryTensorOpWeight ( Scale layer in TRT ). But there are cases that this conversion might fail (transpose, broadcasting on batch, etc), which BinaryTensorOpTensor could catch.

But this is only applicable in TRT 4.0 thanks to the Constant layer added in TRT 4.0.

aaroey replied:
Would you please add that as comment? Also, how about:
```
Status status;
if (input 0 is tensor and input 1 is weight) {
  status = BinaryTensorOpWeight(..., false, ...);
} else if (input 0 is weight and input 1 is tensor) {
  status = BinaryTensorOpWeight(..., true, ...);

#if NV_TENSORRT_MAJOR == 3
} else {
#else
}
if ((input 0 is tensor and input 1 is tensor) || !status.ok()) {
#endif

  status = BinaryTensorOpTensor(...);
}
#endif

return status;
```

jjsjann123 replied:
Looks good. Will do.

jjsjann123 committed:
addressing review comments

jjsjann123 committed:
enabling shape inference function after cherry picking the fix

jjsjann123 committed:
merging format/style fixed

aaroey commented:
Thanks for the fix. The c++ changes looks good to me. I read the python tests and they have duplicate logic comparing to tf_trt_integration_test.py. I would prefer to remove them in this PR and put them in a new PR, and I can merge them with tf_trt_integration_test.py. What do you think?

aaroey commented on tensorflow/contrib/tensorrt/test/tf_trt_integration_test.py:
```diff
@@ -124,15 +124,15 @@ def GetMultiEngineGraphDef(dtype=dtypes.float32):
-            gdef=GetSingleEngineGraphDef(),
-            num_expected_engines=1,
-            expected_output_dims=(100, 6, 6, 6)),
+    TestGraph(
```
Please revert the changes to this file.

jjsjann123 committed:
removing unit tests (for follow up PR)
disabling shape inference function (Because of the fully defined shape requirement from TF)

aaroey commented:
Thanks so much for the patient, LGTM.
